File: server/storage.ts
import { type User, type InsertUser, type Visitor, type InsertVisitor, type VisitorCounts, users, visitors } from "@shared/schema";
import { randomUUID } from "crypto";
import { promises as fs } from "fs";
import path from "path";
import { drizzle } from "drizzle-orm/neon-http";
import { neon } from "@neondatabase/serverless";
import { eq, gte, and, sql } from "drizzle-orm";
import { MongoClient, type Db } from "mongodb";

export interface IStorage {
  getUser(id: string): Promise<User | undefined>;
  getUserByUsername(username: string): Promise<User | undefined>;
  createUser(user: InsertUser): Promise<User>;
  
  // Visitor tracking methods
  trackVisitor(visitor: InsertVisitor): Promise<Visitor>;
  getVisitorCounts(period?: 'today' | 'week' | 'month' | 'year'): Promise<VisitorCounts | Partial<VisitorCounts>>;
  getAllVisitors(): Promise<Visitor[]>;
}

export class JsonFileStorage implements IStorage {
  private dataFilePath: string;
  private data: {
    users: Map<string, User>;
    visitors: Map<string, Visitor>;
  };

  constructor(dataFilePath: string = "data/visitors_data.json") {
    this.dataFilePath = dataFilePath;
    this.data = {
      users: new Map(),
      visitors: new Map(),
    };
    this.loadData();
  }

  private async ensureDataDirectory(): Promise<void> {
    const dir = path.dirname(this.dataFilePath);
    try {
      await fs.mkdir(dir, { recursive: true });
    } catch (error) {
      // Directory might already exist, ignore error
    }
  }

  private async loadData(): Promise<void> {
    try {
      await this.ensureDataDirectory();
      const data = await fs.readFile(this.dataFilePath, "utf-8");
      const parsedData = JSON.parse(data);
      
      // Convert arrays back to Maps
      this.data.users = new Map(parsedData.users || []);
      this.data.visitors = new Map(parsedData.visitors || []);
    } catch (error) {
      // File doesn't exist or is invalid, start with empty data
      this.data = {
        users: new Map(),
        visitors: new Map(),
      };
      await this.saveData();
    }
  }

  private async saveData(): Promise<void> {
    try {
      await this.ensureDataDirectory();
      const dataToSave = {
        users: Array.from(this.data.users.entries()),
        visitors: Array.from(this.data.visitors.entries()),
      };
      await fs.writeFile(this.dataFilePath, JSON.stringify(dataToSave, null, 2));
    } catch (error) {
      console.error("Error saving data to JSON file:", error);
    }
  }

  async getUser(id: string): Promise<User | undefined> {
    return this.data.users.get(id);
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    return Array.from(this.data.users.values()).find(
      (user) => user.username === username,
    );
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    const id = randomUUID();
    const user: User = { ...insertUser, id };
    this.data.users.set(id, user);
    await this.saveData();
    return user;
  }

  async trackVisitor(insertVisitor: InsertVisitor): Promise<Visitor> {
    const id = randomUUID();
    const visitor: Visitor = {
      userId: insertVisitor.userId || null,
      page: insertVisitor.page,
      referrer: insertVisitor.referrer || null,
      userAgent: insertVisitor.userAgent || null,
      ip: insertVisitor.ip || null,
      id,
      timestamp: new Date(),
    };
    this.data.visitors.set(id, visitor);
    await this.saveData();
    return visitor;
  }

  async getAllVisitors(): Promise<Visitor[]> {
    return Array.from(this.data.visitors.values());
  }

  async getVisitorCounts(period?: 'today' | 'week' | 'month' | 'year'): Promise<VisitorCounts | Partial<VisitorCounts>> {
    const now = new Date();
    const visitors = Array.from(this.data.visitors.values());
    
    const startOfToday = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const startOfWeek = new Date(now);
    startOfWeek.setDate(now.getDate() - now.getDay());
    startOfWeek.setHours(0, 0, 0, 0);
    
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    const startOfYear = new Date(now.getFullYear(), 0, 1);

    const counts = {
      today: visitors.filter(v => v.timestamp && new Date(v.timestamp) >= startOfToday).length,
      week: visitors.filter(v => v.timestamp && new Date(v.timestamp) >= startOfWeek).length,
      month: visitors.filter(v => v.timestamp && new Date(v.timestamp) >= startOfMonth).length,
      year: visitors.filter(v => v.timestamp && new Date(v.timestamp) >= startOfYear).length,
    };

    if (period) {
      return { [period]: counts[period] };
    }

    return counts;
  }
}

export class MemStorage implements IStorage {
  private users: Map<string, User>;
  private visitors: Map<string, Visitor>;

  constructor() {
    this.users = new Map();
    this.visitors = new Map();
  }

  async getUser(id: string): Promise<User | undefined> {
    return this.users.get(id);
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    return Array.from(this.users.values()).find(
      (user) => user.username === username,
    );
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    const id = randomUUID();
    const user: User = { ...insertUser, id };
    this.users.set(id, user);
    return user;
  }

  async trackVisitor(insertVisitor: InsertVisitor): Promise<Visitor> {
    const id = randomUUID();
    const visitor: Visitor = {
      userId: insertVisitor.userId || null,
      page: insertVisitor.page,
      referrer: insertVisitor.referrer || null,
      userAgent: insertVisitor.userAgent || null,
      ip: insertVisitor.ip || null,
      id,
      timestamp: new Date(),
    };
    this.visitors.set(id, visitor);
    return visitor;
  }

  async getAllVisitors(): Promise<Visitor[]> {
    return Array.from(this.visitors.values());
  }

  async getVisitorCounts(period?: 'today' | 'week' | 'month' | 'year'): Promise<VisitorCounts | Partial<VisitorCounts>> {
    const now = new Date();
    const visitors = Array.from(this.visitors.values());
    
    const startOfToday = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const startOfWeek = new Date(now);
    startOfWeek.setDate(now.getDate() - now.getDay());
    startOfWeek.setHours(0, 0, 0, 0);
    
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    const startOfYear = new Date(now.getFullYear(), 0, 1);

    const counts = {
      today: visitors.filter(v => v.timestamp && v.timestamp >= startOfToday).length,
      week: visitors.filter(v => v.timestamp && v.timestamp >= startOfWeek).length,
      month: visitors.filter(v => v.timestamp && v.timestamp >= startOfMonth).length,
      year: visitors.filter(v => v.timestamp && v.timestamp >= startOfYear).length,
    };

    if (period) {
      return { [period]: counts[period] };
    }

    return counts;
  }
}

export class DatabaseStorage implements IStorage {
  private db;

  constructor() {
    if (!process.env.DATABASE_URL) {
      throw new Error("DATABASE_URL environment variable is required");
    }
    const sql = neon(process.env.DATABASE_URL);
    this.db = drizzle(sql);
  }

  async getUser(id: string): Promise<User | undefined> {
    const result = await this.db.select().from(users).where(eq(users.id, id)).limit(1);
    return result[0];
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    const result = await this.db.select().from(users).where(eq(users.username, username)).limit(1);
    return result[0];
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    const result = await this.db.insert(users).values(insertUser).returning();
    return result[0];
  }

  async trackVisitor(insertVisitor: InsertVisitor): Promise<Visitor> {
    const result = await this.db.insert(visitors).values(insertVisitor).returning();
    return result[0];
  }

  async getAllVisitors(): Promise<Visitor[]> {
    return await this.db.select().from(visitors);
  }

  async getVisitorCounts(period?: 'today' | 'week' | 'month' | 'year'): Promise<VisitorCounts | Partial<VisitorCounts>> {
    const now = new Date();
    
    const startOfToday = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const startOfWeek = new Date(now);
    startOfWeek.setDate(now.getDate() - now.getDay());
    startOfWeek.setHours(0, 0, 0, 0);
    
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    const startOfYear = new Date(now.getFullYear(), 0, 1);

    if (period) {
      let startDate: Date;
      switch (period) {
        case 'today': startDate = startOfToday; break;
        case 'week': startDate = startOfWeek; break;
        case 'month': startDate = startOfMonth; break;
        case 'year': startDate = startOfYear; break;
      }
      
      const result = await this.db.select({ count: sql<number>`count(*)::int` })
        .from(visitors)
        .where(gte(visitors.timestamp, startDate));
      
      return { [period]: result[0].count };
    }

    // Get all counts in parallel using SQL
    const [todayResult, weekResult, monthResult, yearResult] = await Promise.all([
      this.db.select({ count: sql<number>`count(*)::int` }).from(visitors).where(gte(visitors.timestamp, startOfToday)),
      this.db.select({ count: sql<number>`count(*)::int` }).from(visitors).where(gte(visitors.timestamp, startOfWeek)),
      this.db.select({ count: sql<number>`count(*)::int` }).from(visitors).where(gte(visitors.timestamp, startOfMonth)),
      this.db.select({ count: sql<number>`count(*)::int` }).from(visitors).where(gte(visitors.timestamp, startOfYear)),
    ]);

    return {
      today: todayResult[0].count,
      week: weekResult[0].count,
      month: monthResult[0].count,
      year: yearResult[0].count,
    };
  }
}

export class MongoDBStorage implements IStorage {
  private client: MongoClient;
  private db: Db;
  private initialized: boolean = false;

  constructor(connectionString: string) {
    this.client = new MongoClient(connectionString);
    this.db = this.client.db("visitors_analytics");
  }

  async initialize(): Promise<void> {
    if (!this.initialized) {
      await this.client.connect();
      this.initialized = true;
      console.log("Connected to MongoDB successfully");
    }
  }

  async getUser(id: string): Promise<User | undefined> {
    await this.initialize();
    const user = await this.db.collection("users").findOne({ id });
    if (!user) return undefined;
    return {
      id: user.id,
      username: user.username,
      password: user.password,
    };
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    await this.initialize();
    const user = await this.db.collection("users").findOne({ username });
    if (!user) return undefined;
    return {
      id: user.id,
      username: user.username,
      password: user.password,
    };
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    await this.initialize();
    const id = randomUUID();
    const user: User = { ...insertUser, id };
    await this.db.collection("users").insertOne(user);
    return user;
  }

  async trackVisitor(insertVisitor: InsertVisitor): Promise<Visitor> {
    await this.initialize();
    const id = randomUUID();
    const visitor: Visitor = {
      id,
      userId: insertVisitor.userId || null,
      page: insertVisitor.page,
      referrer: insertVisitor.referrer || null,
      userAgent: insertVisitor.userAgent || null,
      ip: insertVisitor.ip || null,
      timestamp: new Date(),
    };
    await this.db.collection("visitors").insertOne(visitor);
    return visitor;
  }

  async getAllVisitors(): Promise<Visitor[]> {
    await this.initialize();
    const visitors = await this.db.collection("visitors").find({}).toArray();
    return visitors.map(v => ({
      id: v.id,
      userId: v.userId,
      page: v.page,
      referrer: v.referrer,
      userAgent: v.userAgent,
      ip: v.ip,
      timestamp: v.timestamp,
    }));
  }

  async getVisitorCounts(period?: 'today' | 'week' | 'month' | 'year'): Promise<VisitorCounts | Partial<VisitorCounts>> {
    await this.initialize();
    const now = new Date();
    
    const startOfToday = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const startOfWeek = new Date(now);
    startOfWeek.setDate(now.getDate() - now.getDay());
    startOfWeek.setHours(0, 0, 0, 0);
    
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    const startOfYear = new Date(now.getFullYear(), 0, 1);

    if (period) {
      let startDate: Date;
      switch (period) {
        case 'today': startDate = startOfToday; break;
        case 'week': startDate = startOfWeek; break;
        case 'month': startDate = startOfMonth; break;
        case 'year': startDate = startOfYear; break;
      }
      
      const count = await this.db.collection("visitors").countDocuments({
        timestamp: { $gte: startDate }
      });
      
      return { [period]: count };
    }

    // Get all counts in parallel
    const [todayCount, weekCount, monthCount, yearCount] = await Promise.all([
      this.db.collection("visitors").countDocuments({ timestamp: { $gte: startOfToday } }),
      this.db.collection("visitors").countDocuments({ timestamp: { $gte: startOfWeek } }),
      this.db.collection("visitors").countDocuments({ timestamp: { $gte: startOfMonth } }),
      this.db.collection("visitors").countDocuments({ timestamp: { $gte: startOfYear } }),
    ]);

    return {
      today: todayCount,
      week: weekCount,
      month: monthCount,
      year: yearCount,
    };
  }
}

// Use MongoDB if MONGODB_URI is set, otherwise use DatabaseStorage if DATABASE_URL is set, otherwise fall back to JsonFileStorage
export const storage = process.env.MONGODB_URI
  ? new MongoDBStorage(process.env.MONGODB_URI)
  : process.env.DATABASE_URL 
    ? new DatabaseStorage() 
    : new JsonFileStorage();

File: api/storage.py
import os
import json
from datetime import datetime, timedelta
from typing import Optional, Dict, List, Union
from uuid import uuid4
from pymongo import MongoClient
from abc import ABC, abstractmethod

class IStorage(ABC):
    @abstractmethod
    def get_user(self, user_id: str) -> Optional[Dict]:
        pass
    
    @abstractmethod
    def get_user_by_username(self, username: str) -> Optional[Dict]:
        pass
    
    @abstractmethod
    def create_user(self, user_data: Dict) -> Dict:
        pass
    
    @abstractmethod
    def track_visitor(self, visitor_data: Dict) -> Dict:
        pass
    
    @abstractmethod
    def get_visitor_counts(self, period: Optional[str] = None) -> Dict:
        pass
    
    @abstractmethod
    def get_all_visitors(self) -> List[Dict]:
        pass

class JSONFileStorage(IStorage):
    def __init__(self, data_file_path: str = "data/visitors_data.json"):
        self.data_file_path = data_file_path
        self.data = {"users": [], "visitors": []}
        self._load_data()
    
    def _ensure_data_directory(self):
        os.makedirs(os.path.dirname(self.data_file_path), exist_ok=True)
    
    def _load_data(self):
        try:
            self._ensure_data_directory()
            with open(self.data_file_path, 'r') as f:
                self.data = json.load(f)
                if "users" not in self.data:
                    self.data["users"] = []
                if "visitors" not in self.data:
                    self.data["visitors"] = []
        except (FileNotFoundError, json.JSONDecodeError):
            self.data = {"users": [], "visitors": []}
            self._save_data()
    
    def _save_data(self):
        try:
            self._ensure_data_directory()
            with open(self.data_file_path, 'w') as f:
                json.dump(self.data, f, indent=2, default=str)
        except Exception as e:
            print(f"Error saving data to JSON file: {e}")
    
    def get_user(self, user_id: str) -> Optional[Dict]:
        for user in self.data["users"]:
            if user.get("id") == user_id:
                return user
        return None
    
    def get_user_by_username(self, username: str) -> Optional[Dict]:
        for user in self.data["users"]:
            if user.get("username") == username:
                return user
        return None
    
    def create_user(self, user_data: Dict) -> Dict:
        user = {
            "id": str(uuid4()),
            "username": user_data["username"],
            "password": user_data["password"]
        }
        self.data["users"].append(user)
        self._save_data()
        return user
    
    def track_visitor(self, visitor_data: Dict) -> Dict:
        visitor = {
            "id": str(uuid4()),
            "userId": visitor_data.get("userId"),
            "page": visitor_data["page"],
            "referrer": visitor_data.get("referrer"),
            "userAgent": visitor_data.get("userAgent"),
            "ip": visitor_data.get("ip"),
            "timestamp": datetime.utcnow().isoformat()
        }
        self.data["visitors"].append(visitor)
        self._save_data()
        return visitor
    
    def get_all_visitors(self) -> List[Dict]:
        return self.data["visitors"]
    
    def get_visitor_counts(self, period: Optional[str] = None) -> Dict:
        now = datetime.utcnow()
        
        start_of_today = datetime(now.year, now.month, now.day)
        start_of_week = now - timedelta(days=now.weekday())
        start_of_week = datetime(start_of_week.year, start_of_week.month, start_of_week.day)
        start_of_month = datetime(now.year, now.month, 1)
        start_of_year = datetime(now.year, 1, 1)
        
        visitors = self.data["visitors"]
        
        def count_since(start_date):
            count = 0
            for v in visitors:
                try:
                    timestamp = datetime.fromisoformat(v["timestamp"].replace("Z", "+00:00"))
                    if timestamp >= start_date:
                        count += 1
                except:
                    pass
            return count
        
        counts = {
            "today": count_since(start_of_today),
            "week": count_since(start_of_week),
            "month": count_since(start_of_month),
            "year": count_since(start_of_year)
        }
        
        if period:
            return {period: counts.get(period, 0)}
        
        return counts

class MongoDBStorage(IStorage):
    def __init__(self, connection_string: str):
        self.client = MongoClient(connection_string)
        self.db = self.client["visitors_analytics"]
        self.initialized = False
    
    def initialize(self):
        if not self.initialized:
            self.client.admin.command('ping')
            self.initialized = True
            print("Connected to MongoDB successfully")
    
    def get_user(self, user_id: str) -> Optional[Dict]:
        self.initialize()
        user = self.db.users.find_one({"id": user_id})
        if user:
            user.pop("_id", None)
        return user
    
    def get_user_by_username(self, username: str) -> Optional[Dict]:
        self.initialize()
        user = self.db.users.find_one({"username": username})
        if user:
            user.pop("_id", None)
        return user
    
    def create_user(self, user_data: Dict) -> Dict:
        self.initialize()
        user = {
            "id": str(uuid4()),
            "username": user_data["username"],
            "password": user_data["password"]
        }
        self.db.users.insert_one(user.copy())
        return user
    
    def track_visitor(self, visitor_data: Dict) -> Dict:
        self.initialize()
        visitor = {
            "id": str(uuid4()),
            "userId": visitor_data.get("userId"),
            "page": visitor_data["page"],
            "referrer": visitor_data.get("referrer"),
            "userAgent": visitor_data.get("userAgent"),
            "ip": visitor_data.get("ip"),
            "timestamp": datetime.utcnow()
        }
        self.db.visitors.insert_one(visitor.copy())
        return visitor
    
    def get_all_visitors(self) -> List[Dict]:
        self.initialize()
        visitors = list(self.db.visitors.find({}))
        for v in visitors:
            v.pop("_id", None)
        return visitors
    
    def get_visitor_counts(self, period: Optional[str] = None) -> Dict:
        self.initialize()
        now = datetime.utcnow()
        
        start_of_today = datetime(now.year, now.month, now.day)
        start_of_week = now - timedelta(days=now.weekday())
        start_of_week = datetime(start_of_week.year, start_of_week.month, start_of_week.day)
        start_of_month = datetime(now.year, now.month, 1)
        start_of_year = datetime(now.year, 1, 1)
        
        if period:
            start_dates = {
                "today": start_of_today,
                "week": start_of_week,
                "month": start_of_month,
                "year": start_of_year
            }
            start_date = start_dates.get(period)
            if start_date:
                count = self.db.visitors.count_documents({"timestamp": {"$gte": start_date}})
                return {period: count}
            return {period: 0}
        
        counts = {
            "today": self.db.visitors.count_documents({"timestamp": {"$gte": start_of_today}}),
            "week": self.db.visitors.count_documents({"timestamp": {"$gte": start_of_week}}),
            "month": self.db.visitors.count_documents({"timestamp": {"$gte": start_of_month}}),
            "year": self.db.visitors.count_documents({"timestamp": {"$gte": start_of_year}})
        }
        
        return counts

def get_storage() -> IStorage:
    mongodb_uri = os.getenv("MONGODB_URI")
    if mongodb_uri:
        return MongoDBStorage(mongodb_uri)
    return JSONFileStorage()

storage = get_storage()
